imx: invalidate cache for specific address range only

iuliana-prodan opened the pull request:
The _memmap_cacheattr_reset linker script variable has
dedicate cache attribute for every 512M in 4GB space
1: write through
2: cache bypass
4: write back
F: invalid access

For imx, implement is_cached() based on
_memmap_cacheattr_reset.

Without this fix, when loading a topology we get:
sof-audio-of 556e8000.dsp: error: reply expected 20 got 12 bytes
sof-audio-of 556e8000.dsp: error: ipc error for 0x30010000 size 20
sof-audio-of 556e8000.dsp: error: DSP failed to add widget id 0 type 23 name : PCM0P stream Passthrough Playback 0 reply 0
sof-audio-of 556e8000.dsp: ASoC: failed to load widget PCM0P
sof-audio-of 556e8000.dsp: ASoC: topology: could not load header: -22
sof-audio-of 556e8000.dsp: error: tplg component load failed -22
sof-audio-of 556e8000.dsp: error: failed to load DSP topology -22
sof-audio-of 556e8000.dsp: ASoC: error at snd_soc_component_probe on 556e8000.dsp: -22
sof-audio-of 556e8000.dsp: ASoC: failed to probe component -22
asoc-simple-card sof-sound-wm8960: ASoC: failed to instantiate card -22
asoc-simple-card: probe of sof-sound-wm8960 failed with error -22

That's because, on mailbox_hostbox_read() we don't invalidate cache
before reading the address.

Fixes: bfe7707f37de ("cavs: disable data cache operations on uncached addresses")
Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan commented:
checkpatch warning is a false positive:
```
WARNING: Possible unwrapped commit description (prefer a maximum 75 chars per line)
#12: 
sof-audio-of 556e8000.dsp: error: DSP failed to add widget id 0 type 23 name : PCM0P stream Passthrough Playback 0 reply 0

total: 0 errors, 1 warnings, 0 checks, 11 lines checked
```
I've put the exact error message I get while running without this fix.

paulstelian97 reviewed (commented):
Addresses between 0x80000000 and 0x9FFFFFFF are the ones that are cached; everything outside that is not.

I'd say something like 

`#define is_cached(address) (((uintptr_t)address & 0xE0000000) == 0x80000000)`

is closer to the truth. We need to check out the correct variant for imx8m config as it actually differs.

Saying that a region is cached, when it actually isn't, does no harm to be fair, but is suboptimal.

EDIT: It's the same for imx8m actually, so you can just put my definition in the CONFIG_IMX conditional.

dbaluta commented:
@iuliana-prodan good catch! Indeed as @paulstelian97 cache attributes for i.MX are set in the linker script file

```
src/platform/imx8/imx8.x.in

_memmap_cacheattr_imx8_wt_allvalid = 0x22212222;                                                                                                                                                           
PROVIDE(_memmap_cacheattr_reset = _memmap_cacheattr_imx8_wt_allvalid);
```

lgirdwood commented:
I wont apply this until NXP all approve, but doing something like the Intel version is also acceptable.

lgirdwood commented:
For rc2.

paulstelian97 commented:
> I wont apply this until NXP all approve, but doing something like the Intel version is also acceptable.

The problem is that we don't have an aliasing of any kind (cache_to_uncache and uncache_to_cache are no-ops), we just have the range from 0x80000000 to 0x9FFFFFFF (which contains all working memory currently used for all heaps) that is cached, while I/O ranges are uncached (which are below 0x80000000). So we just need to flush the caches for the actually-cached ranges, but not flush them for any I/O range.

dbaluta commented:
@lgirdwood the patch that introduced the problem for NXP
-  (bfe7707 ("cavs: disable data cache operations on uncached addresses") 

is not in v1.8.

Do you plan to add it to v1.8?

iuliana-prodan commented:
@dbaluta @paulstelian97 I've updated the pull request.
While here, I've also added information for linker script variable that contains the cache attributes.

dbaluta commented on src/arch/xtensa/include/arch/lib/cache.h:
```diff
@@ -22,7 +22,26 @@
+ * 4: write back
+ * F: invalid access
+ */
+extern uint32_t _memmap_cacheattr_reset;
```
Here we assume that this are all the possible values. We need to check Xtensa manuals and see if there are any other cache options.

lgirdwood replied:
Fix looks good, but pls check the values and this can go in rc2.

paulstelian97 replied:
I have studied the values a bit in the manual, and based on my understanding possible values can be:

0,1,3: Cached, various ways (3 may be unsupported by certain systems, 0 is no-allocate so it may not be useful for generic usage)
4,5: Write back (5 is no-allocate so it may not be useful); may be unsupported according to the manual
2: Bypass
E: Cache isolate, supposedly for testing so it's not of interest to us
F: Invalid access
6-D: Reserved, may not appear currently

In practice 1, 2, 4 and F are the only useful ones so the comment is fine only mentioning them.

dbaluta replied:
@paulstelian97 can you send us a link with documentation? Also, where to find this info. I grepped yesterday through my xtensa docs but couldn't fine this :).

iuliana-prodan replied:
Yes, @paulstelian97 I also looked for this and couldn't find anything useful.
Can you please share with us the doc (link)? Thanks!

paulstelian97 replied:
https://0x04.net/~mwk/doc/xtensa.pdf Section 4.6.3.3 (table 4-104 on page 155 of the manual/179 of the PDF) contains the info. It's the first result on the Google search "xtensa cacheattr register".

lgirdwood commented:
> Do you plan to add it to v1.8?

I was planning to yes, as it fixed a zephyr issue.

paulstelian97 commented on src/arch/xtensa/include/arch/lib/cache.h:
```diff
@@ -22,7 +22,26 @@
+ */
+extern uint32_t _memmap_cacheattr_reset;
+
+#define _addr_range(address) (((uintptr_t)(address) >> 29) & 0xF)
```
AND with 0x7? Only 3 bits are of interest to us, especially since you're shifting 29 bits out (the topmost bit is always zero).

dbaluta replied:
a comment with natural language description of what we are doing here is welcomed.

lgirdwood replied:
Fwiw, the linker script is pre-processed so you could set the ```memmap_cacheattr_reset``` using the same macros for shift sizes/bits that sre used to construct the ```is_cacheable()``` macro.

paulstelian97 commented on src/arch/xtensa/include/arch/lib/cache.h:
```diff
@@ -22,7 +22,26 @@
+#define _addr_range(address) (((uintptr_t)(address) >> 29) & 0xF)
+#define _addr_shift(address) ((_addr_range(address)) << 2)
+#define _addr_attr(address) ((((uint32_t)(&_memmap_cacheattr_reset)) >> \
+			     (_addr_shift(address))) & 0xF)
```
Same comment here.

marc-hb commented:
The QEMU IMX timeout at https://github.com/thesofproject/sof/pull/4177/checks?check_run_id=2575162933 should now be fixed by docker downgrade #4186, please try "re-run jobs" in the "checks" tab.

@keyonjie
